File: algs/src/main/python/NextGreaterElementIII.py
"""
Given a positive integer n, find the smallest integer which has exactly the same
digits existing in the integer n and is greater in value than n.
If no such positive integer exists, return -1.

Note that the returned integer should fit in 32-bit integer,
if there is a valid answer but it does not fit in 32-bit integer, return -1.

Example 1:
Input: n = 12
Output: 21

Example 2:
Input: n = 21
Output: -1

Constraints:
    1 <= n <= 2^31 - 1
"""
def nextGreaterElement(self, n: int) -> int:
    max_n = 2 ** 31 - 1

    n_str = list(str(n))
    i = len(n_str) - 1

    while (i := i - 1) >= 0:
        if n_str[i] >= n_str[i + 1]:
            continue

        j = i + 1

        for k in range(j + 1, len(n_str)):
            if (n_str[k] < n_str[j]) and (n_str[k] > n_str[i]):
                j = k

        n_i = n_str[i]
        n_j = n_str[j]

        n_str[i] = n_j
        n_str[j] = n_i

        m_str = n_str[:i + 1] + sorted(n_str[i + 1:])
        m = int(''.join(m_str))

        return m if m <= max_n else -1

    return -1

File: algs/src/main/java/algs/assignments/leetcode/NextGreaterElementIII.java
package algs.assignments.leetcode;

/**
 * Given a positive 32-bit integer n, you need to find the smallest 32-bit integer which has exactly
 * the same digits existing in the integer n and is greater in value than n. If no such positive 32-bit
 * integer exists, you need to return -1.
 *
 * Example 1:
 * Input: 12
 * Output: 21
 *
 *
 * Example 2:
 * Input: 21
 * Output: -1
 */
public class NextGreaterElementIII {

    public int nextGreaterElement(int n) {
        if (n == Integer.MAX_VALUE)
            return -1;

        String nstr = String.valueOf(n);

        for (int i = nstr.length()-1; i > 0; i--) {
            if (nstr.charAt(i) > nstr.charAt(i-1))
            {
                int a = (nstr.charAt(i-1) - '0')*10 + (nstr.charAt(i) - '0');
                int b = (nstr.charAt(i) - '0')*10 + (nstr.charAt(i-1) - '0');

                if (Integer.MAX_VALUE - n > b-a)
                    return Integer.parseInt(nstr.substring(0, i-1) + nstr.charAt(i) + nstr.charAt(i-1) + nstr.substring(i+1));
                else
                    return -1;
            }
        }

        return -1;
    }

}
